Google.Api.Gax.Grpc/Rest/RpcCancellationContext.cs

```diff
@@ -0,0 +1,153 @@
+        string rpcName,
+        CancellationTokenSource deadlineCancellationTokenSource,
+        CancellationToken callOptionsCancellationToken)
+    {
+        _rpcName = rpcName;
+        _deadlineCancellationTokenSource = deadlineCancellationTokenSource;
+        var deadlineToken = _deadlineCancellationTokenSource?.Token ?? default;
+        _overallCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(callOptionsCancellationToken, deadlineToken);
+    }
+
+    /// <summary>
+    /// Creates an instance using the given cancellation token source for deadlines and the given
+    /// call cancellation token. This is only present for test purposes.
+    /// </summary>
+    internal static RpcCancellationContext ForTesting(
+        string rpcName,
+        CancellationTokenSource deadlineCancellationTokenSource,
+        CancellationToken callCancellationToken) =>
+        new RpcCancellationContext(rpcName, deadlineCancellationTokenSource, callCancellationToken);
+
+    /// <summary>
+    /// Creates a cancellation context from the given call options.
+    /// </summary>
+    /// <param name="rpcName">The name of the RPC, used to report exceptions.</param>
+    /// <param name="options"></param>
+    /// <exception cref="RpcException">The options contain a deadline that has already elapsed</exception>
+    internal static RpcCancellationContext FromOptions(string rpcName, CallOptions options)
+    {
+        CancellationTokenSource deadlineCancellationTokenSource = null;
+        if (options.Deadline is DateTime deadline)
+        {
+            var timeout = deadline - DateTime.UtcNow;
+            if (timeout.TotalMilliseconds <= 0)
+            {
+                throw new RpcException(new Status(StatusCode.DeadlineExceeded, $"RPC '{rpcName}' timed out"));
+            }
+            deadlineCancellationTokenSource = new CancellationTokenSource(timeout);
+        }
+        return new RpcCancellationContext(rpcName, deadlineCancellationTokenSource, options.CancellationToken);
+    }
+
+    /// <summary>
+    /// Combines <paramref name="cancellationToken"/> with the RPC's cancellation tokens, and waits for
+    /// the task provided by <paramref name="taskProvider"/> to complete, converting any <see cref="OperationCanceledException"/>
+    /// into an <see cref="RpcException"/> with an appropriate status (DeadlineExceeded or Cancelled depending
+    /// on which token was responsible for the original exception).
+    /// </summary>
+    /// <param name="taskProvider">A function which returns a task to await.</param>
+    /// <param name="cancellationToken">An additional cancellation token, defaulting to "no extra cancellation token,
+    /// just use the RPC's cancellation tokens".</param>
+    /// <returns>A task which will complete when the provided task completes.</returns>
+    public async Task AwaitAsync(Func<CancellationToken, Task> taskProvider, CancellationToken cancellationToken = default)
```

Comment:
maybe the name of this method should be `RunAsync` so that callers do `await RunAsync` and not `await AwaitAsync`

Reply:
Yes, good plan. Will do.